src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,388 @@
+/*******************************************************************\
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited.
+
+\*******************************************************************/
+
+#include "replace_function_bodies.h"
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <memory>
+#include <sstream>
+#include <util/make_unique.h>
+#include <util/message.h>
+#include <langapi/language_util.h>
+#include <util/string_utils.h>
+
+/// Base class for replace_function_body implementations
+class replace_function_bodiest
+{
+protected:
+  /// Produce a body for the passed function
+  /// At this point the body of function is always empty,
+  /// and all function parameters have identifiers
+  /// @param function whose body to generate
+  /// @param symbol_table of the current goto program
+  /// @param function_name Identifier of function
+  virtual void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) = 0;
+
+public:
+  virtual ~replace_function_bodiest() = default;
+
+  /// Replace the function body with one based on the implementation
+  /// This will work the same whether or not the function already has a body
+  /// @param function whose body to replace
+  /// @param symbol_table of the current goto program
+  /// @param function_name Identifier of function
+  void replace_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name)
+  {
+    function.body.clear();
+    generate_parameter_names(function, symbol_table, function_name);
+    generate_function_body(function, symbol_table, function_name);
+  }
+
+private:
+  /// Generate parameter names for unnamed parameters.
+  /// CBMC expect functions to have parameter names
+  /// if they are called and have a body
+  void generate_parameter_names(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
+  {
+    namespacet ns(symbol_table);
+    int param_counter = 0;
+    for(auto &parameter : function.type.parameters())
+    {
+      if(parameter.get_identifier().empty())
+      {
+        const std::string param_base_name =
+          parameter.get_base_name().empty()
+            ? "__param$" + std::to_string(param_counter++)
+            : id2string(parameter.get_base_name());
+        irep_idt new_param_identifier =
+          id2string(function_name) + "::" + param_base_name;
+        parameter.set_base_name(param_base_name);
+        parameter.set_identifier(new_param_identifier);
+        parameter_symbolt new_param_sym;
+        new_param_sym.name = new_param_identifier;
+        new_param_sym.type = parameter.type();
+        new_param_sym.base_name = param_base_name;
+        new_param_sym.mode = ID_C;
```

Comment:
No, make that the same mode as the function symbol's mode (which you look up in the next instruction).